addOns/pscanrules/src/test/java/org/zaproxy/zap/extension/pscanrules/payloader/ExtensionPayloaderUnitTest.java

```diff
@@ -0,0 +1,78 @@
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2023 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.pscanrules.payloader;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import org.junit.jupiter.api.Test;
+import org.parosproxy.paros.control.Control;
+import org.parosproxy.paros.extension.ExtensionHook;
+import org.parosproxy.paros.extension.ExtensionLoader;
+import org.parosproxy.paros.model.Model;
+import org.zaproxy.zap.extension.custompayloads.ExtensionCustomPayloads;
+import org.zaproxy.zap.extension.pscanrules.ExtensionPscanRules;
+import org.zaproxy.zap.testutils.TestUtils;
+
+class ExtensionPayloaderUnitTest {
+
+    @Test
+    void shouldHaveAName() {
+        // Given
+        ExtensionPayloader ep = new ExtensionPayloader();
+        TestUtils.mockMessages(new ExtensionPscanRules());
+        // When / Then
+        assertThat(ep.getName(), is(equalTo("ExtensionPayloaderPscanRulesRelease")));
+        assertThat(ep.getUIName(), is(equalTo("Passive Scan Rules Release Custom Payloads")));
+    }
+
+    @Test
+    void shouldBeUnloadable() {
+        // Given
+        ExtensionPayloader ep = new ExtensionPayloader();
+        // When / Then
+        assertThat(ep.canUnload(), is(equalTo(true)));
+    }
+
+    @Test
+    void shouldHandleExpectedCategoriesOnHookAndUnload() {
+        // Given
+        Model model = mock(Model.class);
+        Model.setSingletonForTesting(model);
+        ExtensionLoader extensionLoader = mock(ExtensionLoader.class);
+        Control.initSingletonForTesting(Model.getSingleton(), extensionLoader);
+        ExtensionPayloader ep = new ExtensionPayloader();
+        ExtensionHook eh = mock(ExtensionHook.class);
+        ExtensionCustomPayloads ecp = mock(ExtensionCustomPayloads.class);
+        given(extensionLoader.getExtension(ExtensionCustomPayloads.class)).willReturn(ecp);
+        // When
+        ep.hook(eh);
+        ep.unload();
+        // When / Then
+        verify(ecp, times(2)).addPayloadCategory(any());
+        verify(ecp, times(2)).removePayloadCategory(any());
```

Comment:
Look at using a captor to ensure what was hooked is what's unloaded :wink:

Reply:
Also https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#48 to verify that `*.setPayloadProvider(null)` happens.

Reply:
@thc202 I wasn't able to sort this out. Am I using mockStatic on the rules or the ExtensionHook? All our other "examples" simply dictate the behavior so aren't really testing anything, other than it behaves as we've dictated, and I don't see a way to mock a static void method.

Reply:
I did get the captor part addressed.

Reply:
🎉🎉 I figured it out :grinning: :grinning: